Update CB subscription notebook

jasper-dijkstra opened the pull request:
(1) Update the notebook to run a Crop Biomass subscription and (2) add a new notebook that show how to run a Crop Biomass subscription for multiple fields. The main changes include:

- Using Planet SDK instead of basic HTTP requests
- Also give an example of using the (raster) delivery option
- In the single field notebook data is requested for a new field in Ohio (as it was better for visualisation in the delivery step).
- In the multiple field notebook data is extracted for three fields in Flevoland, the Netherlands.

mlubej commented on jupyter-notebooks/Subscriptions-API/Crop_Biomass_Example.ipynb:
In cell id `85cffb06` you do 

```python

# Extract only the month and day from the 'item_datetime' column
df_growing_season.loc[:, 'Month_Day'] = df_growing_season['item_datetime'].apply(lambda x: x.strftime('%m-%d'))

# Group by 'Year' and 'Month_Day' and compute the median for 'CB.band-1.mean' for each group
df_grouped = df_growing_season.groupby(['Year', 'Month_Day'])['CB.band-1.mean'].median().reset_index()

# Add a pseudo year to 'Month_Day' for plotting
df_grouped['plot_date'] = df_grouped.apply(lambda row: pd.to_datetime(f'2000-{row.Month_Day}'), axis=1)
```

I find the splitting to `month-day` and grouping by multiple keys confusing, so I would suggest the following:

```python
df_growing_season['item_date'] = df_growing_season['item_datetime'].dt.date
df_grouped = df_growing_season.groupby('item_date')['CB.band-1.mean'].median().reset_index()
```

I understand that you use the dummy year for plotting, but in this case you could also do the apply on the series, no need to do it on the whole dataframe, and it could be simplified to:

```python
df_grouped['plot_date'] = df_grouped.item_date.apply(lambda dt: dt.replace(year=2000)
```

Later on, plotting each year's data can also be:

```python
# Plot each year's data
for year, df_year in df_grouped.groupby('Year'):
    ax.plot(df_year['CB.band-1.mean'], color=colors[i], label=year)
```

jasper-dijkstra replied:
Thanks for the suggestions, I implemented them!

mlubej commented on jupyter-notebooks/Subscriptions-API/Crop_Biomass_Example_Multiple_Fields.ipynb:
In cell ID `ea18ba09` when you load the AOI from a file, you take the first feature from the list of geometries (`AOI = geojson_data["features"][0]`), which you shouldn't do for this notebook. 

Also, change `geometry` to `geometries` in `Optionally, you can also load your geometry from a .geojson file`

______________________________________

In cell ID `efe08563` when you're extracting data from the geojson dict, this can be much more easily achieved with `geopandas`, but I guess the user would have to install that additionally? I think `geopandas` is quite a common library.

```python
gdf = gpd.GeoDataFrame.from_features(AOI)
gdf['uuid'] = None
gdf['status'] = 'to_process'

gdf.head()
```

But perhaps you can leave this as is if the library needs to be installed.

__________________________________________


When creating subscriptions in cell ID `7e785da7`, I think we can make things less convoluted by extracting a function for a single geometry. 

```python
def create_single_subscription(row: pd.Series) -> pd.Series:
    pv_source = planetary_variable_source(
            var_type='biomass_proxy',
            var_id="BIOMASS-PROXY_V4.0_10", 
            geometry=row.geometry_json, 
            start_time=dt.datetime(2020, 1, 1),
            end_time=dt.datetime(2023, 12, 31)
        )

    subscription_name = f"CB_{row.name}"
    request = build_request(subscription_name, source=pv_source, delivery=delivery_config_gcp)
        
    try:
        async with Session(auth=auth) as sess:
            cl = SubscriptionsClient(sess)
            subscription_desc = await cl.create_subscription(request)
        row.uuid=subscription_desc['id']
        row.status=subscription_desc['status']
        print(f"{subscription_desc['id']} named {subscription_desc['name']} created")
    except Exception as e:
        print(f"Failed to create subscription for field: '{row.name}', with message: {e}")
        row.uuid="failed"

    return row

for idx, row in df.iterrows(): 
    if row.status == "to_process" or row.status == "failed":
        updated_row = create_single_subscription(row)
        df.loc[idx] = updated_row
    else:
        print(f"Subscription for {row.name} already exists, with uuid: {row.uuid}")

df.head()
```

_______________________________________________

What I also noticed in the notebook is the potentially suboptimal usage of `iterrows`. When iterating, the first argument is the index, and the second one the pd.Series of the row. I usually name them `idx` and `row`. Accessing the items via `df.loc[idx, "uuid"]` can then be replaced with `row.uuid`, this way you don't need to constantly access the dataframe

______________________________________________

in cell ID `02d74e08` it's easier if you expand the outputs

```python
for name, field_df in data_df.groupby("name"):
    field_df['CB.band-1.mean'].plot(ax=ax, kind='line', style='o', label=name, markersize=3)

```

_____________________________________________

when plotting rasters in cell ID `92cb4923` there is a really cool trick how you can plot the subplots without needing to access row and column indices this way:

```python
from itertools import product

dates_to_plot = ["2022-04-01", "2022-06-01", "2022-08-01"]
data_df = data_df.set_index("uuid").sort_values("name")  # set uuid index and sort by name

# create a list of dates and fields combinations
plot_args = list(product(dates_to_plot, data_df.index))

fig, axs = plt.subplots(nrows=len(data_df), ncols=len(dates_to_plot), figsize=(15, 15))

for ax, (date, field) in zip(axs.flatten(), plot_args):
    ...

    file = (
        f"gs://{bucket_name}/{field}/{date.replace('-','/')}/BIOMASS-PROXY_V4.0_10-{date.replace('-','')}T0000_CB.tiff"
    )
    field_name = uuid_to_name[field]

    with rasterio.env.Env(GOOGLE_APPLICATION_CREDENTIALS=GOOGLE_APPLICATION_CREDENTIALS):
        cb_data = rx.open_rasterio(file, mask_and_scale=True, band_as_variable=True)
        cb_data = cb_data.rename_vars({"band_1": "Crop Biomass [-]"})
        img = cb_data["Crop Biomass [-]"].plot(ax=ax, cmap="Greens", vmin=0, vmax=1, add_colorbar=False)

    ax.set_title(f"{date}")
    ax.grid(True)
    ax.axis("equal")
    ax.set_xlabel("")
    ax.set_ylabel("")
    ax.set_frame_on(False)

for j in range(len(data_df)):
    axs[j, 0].set_ylabel(field_name)
    axs[j, 0].yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f"{x:.4f}"))
    axs[j, 1].set_yticklabels([])
    axs[j, 2].set_yticklabels([])


cbar_ax = fig.add_axes([0.15, 0.05, 0.7, 0.02])
cbar = fig.colorbar(img, cax=cbar_ax, orientation="horizontal")
cbar.set_label("Crop Biomass [-]")

plt.tight_layout()
```

The above reduces the nesting and the need for the indexing. BTW, I think you could also avoid the removal of y labels for the second and third column if you put `sharey=True` in the `plt.subplots` function, but I'm not sure what happens with your custom major ticks on the y axis.
```

jasper-dijkstra replied:
Good points! 

Yesterday I had a discussion with @yoann-malbeteau and we agreed that it might indeed be best to use the GeoPandas library. In addition I have removed the JSON dict, so the only option for the user is to provide a GeoJSON.

I agree that it might make sense to create a function for a single subscription. My only concern is that some `async` logic wille be required (the `create_single_subscription` needs to be an async function, that must be called with `asyncio.run()`), and think that it might add a bit of overhead. What do you think?

I like the suggestions for the raster visualisation! But, we decided to drop this bit altogether, since this notebook will focus more on the processing of multiple fields rather than visualising them.

mlubej replied:
>  In addition I have removed the JSON dict, so the only option for the user is to provide a GeoJSON.

Yes, I agree, makes the notebook better, and then you can simply load the geojson to geopandas.

The only issue is then the fact that geometries have to be provided in GEOJSON, and not as shapely geometries, so you have to add another column

```python
import shapely

gdf['geometry_json'] = gdf.geometry.apply(shapely.to_geojson)
```

and then use that in the planet subscription. `shapely` should come with `geopandas`.

> I agree that it might make sense to create a function for a single subscription. My only concern is that some async logic wille be required (the create_single_subscription needs to be an async function, that must be called with asyncio.run()), and think that it might add a bit of overhead. What do you think?

Hmm, I thought that since it's all wrapped up inside the function that the waiting should happen already there, but I don't have that many experience with async stuff, so it's possible it won't work, and making it work would be overhead, so then you can probably leave it as it is.

> I like the suggestions for the raster visualisation! But, we decided to drop this bit altogether, since this notebook will focus more on the processing of multiple fields rather than visualising them.

Also OK!

jasper-dijkstra replied:
> The only issue is then the fact that geometries have to be provided in GEOJSON, and not as shapely geometries, so you have to add another column

I resolved it using the mapping function from Shapely, which converts a POLYGON in the geometry column to a JSON-like dict:

```python
from shapely.geometry import mapping
```

```python
pv_source = planetary_variable_source(
    var_type='biomass_proxy',
    var_id="BIOMASS-PROXY_V4.0_10", 
    geometry=mapping(gdf.loc[row, "geometry"]),  # Convert the shapely geometry to a JSON-like dict
    start_time=start_date,
    end_time=end_date,
)
```

mlubej commented:
I noticed there is also an old PR related to the same notebook in #306 

I guess this PR supersedes #306, so #306 can be closed when this is merged 

closes #306

yoann-malbeteau commented:
Could you create two cells from the code below to make it clear the info is been saved. 

```python
for row, items in gdf.iterrows(): 

    async with Session(auth=auth) as sess:
        cl = SubscriptionsClient(sess)
        subscription_desc = await cl.get_subscription(subscription_id=gdf.loc[row, 'uuid'])
        
    gdf.loc[row, 'status']=subscription_desc['status']
    print(f"{subscription_desc['id']} named {subscription_desc['name']} is {subscription_desc['status']}")

# Save the GeoDataFrame with updated statuses to the existing GeoJSON file
gdf.to_file(fields_geojson, driver='GeoJSON')
```

matthew-ballard commented:
hey this looks really awesome! I really like how thorough it is and how many useful resources are shared with it.  

I do think you could consider combining the two notebooks into one, but I don't think it's critical.  If you don't, maybe they can be grouped into a folder? Small feedback, though, and not critical so just take it as a suggestion.

I added a commit with some minor adjustments to words/capitalization and such. Just two very small comments

`as the AOI variable in the code below (some examples of public features are provided [here](https://www.planet.com/features/?org=1&loc=0,0,2.86z)).
`
does this work? I see org=1 in the URL, not sure if users will get this.

` the valid_percent in [metadata results (time series)](https://static.prod.planet-labs.com/devcenter/dev/rel-pv-eap/docs/subscriptions/pvs-subs/#metadata-results-time-series).`

this has a link to EAP docs that aren't accessible

jasper-dijkstra commented:
Thanks @matthew-ballard
I moved the two notebooks to a `crop_biomass` subfolder.  

> I see org=1 in the URL, not sure if users will get this

I am not sure here either. I get redirected to a login page, but I am not sure if other users can see the same as me after logging in. Is there a way to check this?

> this has a link to EAP docs that aren't accessible

You are right! I removed this link

yoann-malbeteau commented:
> the valid_percent in [metadata results (time series)](https://static.prod.planet-labs.com/devcenter/dev/rel-pv-eap/docs/subscriptions/pvs-subs/#metadata-results-time-series
I think you want the following link: 
https://developers.planet.com/docs/subscriptions/pvs-subs/#data-delivery

yoann-malbeteau reviewed (approved):
After the last two comments on links, I am good to go
